Add Jun's GSoC final blog (#1240)
HSF/hsf.github.io: _gsocblogs/2022/blog_ROOT_JunZhang.md

---
project: ROOT

> title: Exploration of improving the performance of Clang modules in ROOT and CMSSW

Reviewer: I'd keep the title as it was.

In conclusion, we pinpointed the real issue in Clang. Rather than trying to mitigate the problem on the side of ROOT, I've started to talk with Clang modules code owners and plan to continue the work and cooperate with the upstream to further improve modules in ROOT. Hopefully, we can solve this longstanding issue and make Clang better.

## Other work

Except the modules work, I also done a significant contribution to `clang-repl`, which is an upstream version of ROOT's core interpreter. I improved the correctness and reliability of `clang-repl` by fixing lots of bugs. I also added an initial implementation of the code undo feature. Here're some noticeable patches:

* https://reviews.llvm.org/D126781
* https://reviews.llvm.org/D126682
* https://reviews.llvm.org/D128782
* https://reviews.llvm.org/D130420
* https://reviews.llvm.org/D130831

## It's not just about code!

> During this summer, I have learned a lot about coding stuff, and enhanced my knowledge of C++ and compiler. However, I think the most crucial thing I have learned has nothing to do with programming itself.

Reviewer: ```suggestion
During this summer, I have learned a lot about coding stuff, and enhanced my knowledge of C++ and compilers. However, I think the most crucial thing I have learned has nothing to do with programming itself.
```

Due to the huge legacy codebase and the complexity of C++ semantics itself, the knowledge dispersed between various people and teams. In order to solve the problems, instead of sitting at the desk spend all day coding and debugging, we need to look for the right communication channels to extract the knowledge and turn it into an acceptable solution.

> So remember human beings are collective animals and it's always better to work with a team!

Reviewer: ```suggestion
So remember human beings are collective animals and it's always better to work as a team!
```

```
root [0] edm::X;
```

When ROOT sees `edm`, it will immediately pull up all modules containing it, which is totally unnecessary since we only care about `X`. So my project this year is aiming to eliminate the unnecessary lookup for identifiers like a namespace to reduce memory pressure.

## The Work I have done

At the very beginning of the project, I tried to eliminate all namespaces identifiers lookups as it's the most straightforward idea. But we soon ran into some troubles. The problem is that we lost the ability to recognize identifiers inside those nested namespaces, which is a very common case:

```c++
 error: no member named 'MakeTrivialDataFrame' in namespace 'ROOT::RDF'
   auto d_s = ROOT::RDF::MakeTrivialDataFrame(nEvents);
              ~~~~~~~~~~~^
```

This failure looks hard to solve so I was thinking: Why not achieve our goals step by step and only take top-level namespaces into consideration? So I did. The test results look good but when it comes to the performance improvement, Ugh... the results doesn't look very satisfing. The reason is simple, as there're not too many top-level namspaces, so we can't get significant memory decrease.

> We seem to touch some dark corners in ROOT. During that period, I explored many directions but none seems to work. My mentors and I had some deep discussions, and finally, we came to a new approach: If we cannot ignore those namespaces, why not create a new module that contains all namespaces forward declarations? Because the manually generated module just consists of forward declarations, it's cheap to preload, and ROOT can have all essential knowledge about namespaces with it. So here's the [PR](https://github.com/root-project/root/pull/10910). The performance improvement also looks very promising:

Reviewer: Dark corners of ROOT but also the semantics of C++ does not allow us to do much even outside of ROOT.

## Post GSoC

However, it's not enough. I have done some deep profiling about Clang, and found out the bottleneck:

```
% total        Total        Calls  Function
   99.3  444'432'413       74'169  main [4]
   92.3  412'998'941       29'027  llvm::SmallVectorBase::grow_pod(void*, unsigned long, unsigned long) [5]
   89.2  399'211'718        7'870  clang::CompilerInstance::loadModule(clang::SourceLocation, llvm::ArrayRef<std::pair<clang::IdentifierInfo*, clang::SourceLocation> >, clang::Module::NameVisibilityKind, bool) [6]
   89.2  399'058'838        7'644  clang::ASTReader::ReadAST(llvm::StringRef, clang::serialization::ModuleKind, clang::SourceLocation, unsigned int, llvm::SmallVectorImpl<clang::ASTReader::ImportedSubmodule>*) [7]
   88.8  397'250'710        8'023  cling::Interpreter::loadModule(clang::Module*, bool) [8]
   88.7  397'152'678        7'901  clang::Sema::ActOnModuleImport(clang::SourceLocation, clang::SourceLocation, clang::SourceLocation, llvm::ArrayRef<std::pair<clang::IdentifierInfo*, clang::SourceLocation> >) [9]
   84.0  376'053'039       12'354  TRint::TRint(char const*, int*, char**, void*, int, bool, bool) [10]

```

> From the above, we can see the hottest method is `SmallVectorBase::grow_pod`. That is actually caused by a problematic model: Clang always preallocates a huge buffer that is enough to contain all `SourceLocation` in a module, even without use.

Reviewer: Mention the SparseVector tests.

> To further improve modules in ROOT, I plan to continue the work and cooperate with the upstream, and hopefully, we can solve this longstanding issue and make Clang better.

Reviewer: We should also mentioned that we pinpointed a real issue in clang and started discussions with the relevant people how to properly solve it rather than trying to hide/mitigate the problem on the side of Cling/ROOT.

## Other work

> Except the modules work, I also done an enormous contribution to `clang-repl`, which is an upstream version of ROOT's core interpreter. I improved the correctness and reliability of `clang-repl` by fixing lots of bugs. I also added an initial implementation of the code undo feature. Here're some noticeable patches:

Reviewer: ```suggestion
Except the modules work, I also done a significant contribution to `clang-repl`, which is an upstream version of ROOT's core interpreter. I improved the correctness and reliability of `clang-repl` by fixing lots of bugs. I also added an initial implementation of the code undo feature. Here're some noticeable patches:
```

* https://reviews.llvm.org/D126781
* https://reviews.llvm.org/D126682
* https://reviews.llvm.org/D128782
* https://reviews.llvm.org/D130420
* https://reviews.llvm.org/D130831

## It's not just about code!

> During this summer, I have learned a lot about coding stuff, and enhanced my knowledge of C++ and compiler. However, I think the most crucial thing I have learned has nothing to do with programming itself. In the last two months, I realized the importance of communication. That said, you can't just always code and debug, you need to understand your real goal, and that's achieved by communicating with your teammates. I also realized that my weakness -- not being meticulous enough. That really comes to bite me and makes me make some really stupid mistakes.

Reviewer: > During this summer, I have learned a lot about coding stuff, and enhanced my knowledge of C++ and compiler. However, I think the most crucial thing I have learned has nothing to do with programming itself. In the last two months, I realized the importance of communication. That said, you can't just always code and debug, you need to understand your real goal, and that's achieved by communicating with your teammates.

That happens often when the codebases are huge and the knowledge is dispersed between various people/teams. Then that becomes more of finding the right communication channels to extract the knowledge and turn it into an acceptable solution. So fair amount of social engineering. I would turn this paragraph a bit around saying something around that rather than a self criticism.

> I also realized that my weakness -- not being meticulous enough. That really comes to bite me and makes me make some really stupid mistakes.

I think we should probably rework or drop these.
